libkbfs/dirty_file.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+	"sync"
+)
+
+// syncBlockState represents that state of a block with respect to
+// whether it's dirty or currently being synced.  There can be six states:
+//  1) Not dirty or being synced
+//  2) Dirty and awaiting a sync
+//  3) Being synced and not yet re-dirtied: any write needs to
+//     make a copy of the block before dirtying it.  Also, all writes must be
+//     deferred.
+//  4) Being synced and already re-dirtied: no copies are needed, but all
+//     writes must still be deferred.
+//  5) Finished syncing, and not yet re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+//  5) Finished syncing, and already re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+type syncBlockState int
+
+const (
+	blockNotDirty syncBlockState = iota
+	blockDirty
+	blockSyncingNotDirty
+	blockSyncingAndDirty
+	blockSyncedAndDirty
+	blockSyncedNotDirty
```

Comment:
After reading through this file, it looks like many of the decisions are made on one of `dirty` or `syncing`, and rarely both. Wouldn't it be better to split those out into two states? Or do you prefer to have the states together in order to limit the allowed matrix?